### PR TITLE
refactor(Scripts/AI) ZG20: Venoxis (Snake Boss) modernization

### PR DESCRIPTION
**Pull request by kjack9**
## Changes Proposed:

This PR proposes changes to:
-  [ ] Core (units, players, creatures, game systems).
-  [X] Scripts (bosses, spell scripts, creature scripts).
-  [ ] Database (SAI, creatures, etc).

Modernize ZG20/Venoxis encounter. I did not attempt to increase accuracy in this PR, only to convert to the latest AI conventions. All functionality should remain the same.

## SOURCE:

The changes have been validated through:
- [ ] Live research (checked on live servers, e.g Classic WotLK, Retail, etc.)
- [ ] Sniffs (remember to share them with the open source community!)
- [ ] Video evidence, knowledge databases or other public sources (e.g forums, Wowhead, etc.)
- [ ] The changes promoted by this pull request come partially or entirely from another project (cherry-pick). **Cherry-picks must be committed using the proper --author tag in order to be accepted, thus crediting the original authors, unless otherwise unable to be found**

No changes to AI logic.

## Tests Performed:

This PR has been:
- [X] Tested in-game by the author.
- [ ] Tested in-game by other community members/someone else other than the author/has been live on production servers.
- [ ] This pull request requires further testing and may have edge cases to be tested.

## How to Test the Changes:

- [ ] This pull request can be tested by following the reproduction steps provided in the linked issue
- [ ] This pull request requires further testing. Provide steps to test your changes. If it requires any specific setup e.g multiple players please specify it as well.

1. `.go xyz -12007.214844 -1691.049683 32.284668 309 3.726766`
2. Clear trash
3. Engage boss
4. Observe same mechanics as before this PR

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_venoxis.cpp [hunk 1]
```diff
@@ -80,275 +65,213 @@ enum Phases
...
+    }
 
-    private:
-        uint8 _inMeleeRange;
-        bool _transformed;
-        bool _frenzied;
-    };
+private:
+    // Spawns the Razzashi Cobra adds before the encounter starts
+    void _spawnCobras()
```
Comment (elthehablo) [inline src/server/scripts/EasternKingdoms/ZulGurub/boss_venoxis.cpp:188]: Consider pre spawning them and making them minions. See magtheridon and fathomlord for an example. You need to alter instance_zulgurub for it

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_venoxis.cpp [hunk 2]
```diff
@@ -80,275 +65,213 @@ enum Phases
...
+        });
 
-        void Reset()
+        if (me->ToTempSummon() &&
+            me->ToTempSummon()->GetSummoner() &&
+            me->ToTempSummon()->GetSummoner()->ToCreature()
+        )
         {
-            events.Reset();
+            _venoxis = me->ToTempSummon()->GetSummoner()->ToCreature();
```
Comment (elthehablo) [inline src/server/scripts/EasternKingdoms/ZulGurub/boss_venoxis.cpp:224]: Then you would need creaturedata and GetCreature here too

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_venoxis.cpp [hunk 3]
```diff
@@ -80,275 +65,213 @@ enum Phases
...
+            BossAI::JustSummoned(me->SummonCreature(NPC_RAZZASHI_COBRA, SpawnCobras[i], TEMPSUMMON_CORPSE_DESPAWN));
+        }
+    }
 
-    CreatureAI* GetAI(Creature* creature) const override
+    // Sets all Razzashi Cobras in combat with the zone
+    void _setCobrasInCombat()
     {
-        return GetZulGurubAI<boss_venoxisAI>(creature);
+        BossAI::summons.DoForAllSummons([&](WorldObject* cobra)
```
Comment (elthehablo) [inline src/server/scripts/EasternKingdoms/ZulGurub/boss_venoxis.cpp:199]: Luckily we then have DoForAllMinions. See magtheridon for an example

**Review by elthehablo — CHANGES_REQUESTED** [review]

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_venoxis.cpp (hunk 1 above)
Comment (kjack9) [reply src/server/scripts/EasternKingdoms/ZulGurub/boss_venoxis.cpp:188]: A new system to learn! There's always one more...

Just for my education, what is the functional difference between tracking them as minions and tracking them as summons?

Comment (elthehablo) [reply src/server/scripts/EasternKingdoms/ZulGurub/boss_venoxis.cpp:188]: Mostly that summoning creatures normally is not really blizzlike. In wotlk classic there is strong evidence that at least for boss fights all potential summons are done with spells. Other than that it is useful to prespawn them and put them in the creature table so they can be more easily affected by instance related effects. Plus it is more elegant and cleans up the code imo

**Comment by Nyeriah** [comment]
Consider moving this add to SAI, as it seems it doesn't do anything but cast a single spell. As for pulling with the boss, there are other means to achieve this, formations for once

**Comment by kjack9** [comment]
I have a doubt.

Looks like Razzashi Cobra (11373) is currently setup for both its script AND has SAI assigned.

image

Which makes me wonder why it was moved to a script to begin with. I know there was an `_inMeleeRange` counter that I removed in this PR since it wasn't in use in the script... but now I want to know what mechanic that was for, and if maybe that makes it fall outside the scope of SAI.

I can't find any strategy guide or anything online that indicates that Venoxis' range from his adds is important, but I will keep looking. Does anyone happen to have some insider knowledge on what's going on here?

**Comment by Nyeriah** [comment]
Their script was changed here https://github.com/azerothcore/azerothcore-wotlk/pull/8141 (that eventually got merged despite my change requests - that are pretty much exactly what we are bringing up here)

**Comment by kjack9** [comment]
> Their script was changed here #8141 (that eventually got merged despite my change requests - that are pretty much exactly what we are bringing up here)

Got it, that helps me understand history. Can I safely assume that "All minions from bosses should have a C++ script inside boss script not a SAI or database work." is no longer a rule we are following?

**Comment by Nyeriah** [comment]
> All minions from bosses should have a C++ script inside boss script not a SAI or database work

That was never a rule, that statement didn't come from a maintainer

**Comment by elthehablo** [comment]
Who said that?

**Comment by elthehablo** [comment]
Yes. The whipping post for whomever said that

**Comment by kjack9** [comment]
I removed the custom script for the cobras, cleaned up the SAI, and added entries to `creature_formations` and `linked_respawn`.

Note that if some (not all) of the cobras are killed along with the boss, and then the encounter is reset, all four cobras will respawn. This is remedied by #17597 (tested).

**Review by elthehablo — APPROVED** [review]

**Pull request merged**